### PR TITLE
verify previous path state before backup

### PR DESCRIPTION
**Pull request by ryanfkeepers**
Drive previous paths have a prior bug or condition that allows multiple previousPath metadata folder ID entries within a single drive point to the same previous path string.
In these cases we need to treat the backup as having no prior metadata, and re-run the full delta query to correct the bad state.

---

#### Does this PR need a docs update or release note?

- [x] :white_check_mark: Yes, it's included

#### Type of change

- [x] :bug: Bugfix

#### Test Plan

- [x] :zap: Unit test
- [x] :green_heart: E2E

### CODE REVIEW

File: src/internal/m365/collection/drive/collections.go [hunk 1]
```diff
@@ -72,6 +72,70 @@ func NewCollections(
 	}
 }
 
+func deserializeAndValidateMetadata(
+	ctx context.Context,
+	cols []data.RestoreCollection,
+) (map[string]string, map[string]map[string]string, bool, error) {
+	deltas, prevs, canUse, err := DeserializeMetadata(ctx, cols)
+	if err != nil || !canUse {
+		return deltas, prevs, canUse, clues.Stack(err).OrNil()
```
Comment (ashmrtn) [inline src/internal/m365/collection/drive/collections.go:81]: should we always return false for `canUse`? Or are we also guaranteeing that the caller will always exit if err != nil?

File: src/internal/m365/collection/drive/collections.go [hunk 2]
```diff
@@ -72,6 +72,70 @@ func NewCollections(
...
+	// don't accidentally try to pull in delta results when we should have
+	// enumerated everything instead.
+	//
+	// Loop over the set of previous deltas because it's alright to have paths
+	// without a delta but not to have a delta without paths. This way ensures
+	// we check at least all the path sets for the deltas we have.
+	for drive := range deltas {
+		paths := prevs[drive]
+		if len(paths) == 0 {
+			delete(deltas, drive)
```
Comment (ashmrtn) [inline src/internal/m365/collection/drive/collections.go:95]: nit: we should probably add log messages when we drop stuff so that we can track this stuff down later if needed

File: src/internal/m365/collection/drive/collections.go [hunk 3]
```diff
@@ -72,6 +72,70 @@ func NewCollections(
...
+
+	for driveID, folders := range prevs {
+		for fid, prev := range folders {
+			if otherID, collision := prevPathCollisions[prev]; collision {
+				logger.Ctx(ctx).
+					With(
+						"collision_folder_id_1", fid,
+						"collision_folder_id_2", otherID,
+						"collision_drive_id", driveID,
+						"collision_prev_path", prev).
```
Comment (ashmrtn) [inline src/internal/m365/collection/drive/collections.go:120]: nit: directory path will have PII
```suggestion
						"collision_prev_path", path.LoggableDir(prev)).
```

File: src/internal/m365/collection/drive/collections.go [hunk 4]
```diff
@@ -72,6 +72,70 @@ func NewCollections(
...
+					Info("duplicate previous paths across different folder IDs")
+
+				// If a previous path collision occurs, we need to force a full backup
```
Comment (ashmrtn) [inline src/internal/m365/collection/drive/collections.go:123]:
```suggestion
				// If a previous path collision occurs, we need to force full enumeration
```

File: src/internal/m365/collection/drive/collections.go [hunk 5]
```diff
@@ -72,6 +72,70 @@ func NewCollections(
...
+			if len(prevPath) == 0 {
+				delete(deltas, drive)
+				break
+			}
+		}
+	}
+
+	prevPathCollisions := map[string]string{}
+
+	for driveID, folders := range prevs {
```
Comment (ashmrtn) [inline src/internal/m365/collection/drive/collections.go:112]: Could you please rearrange this? I find it confusing to initialize/clear the map this way instead of just initializing it in the first for-loop
```suggestion
	for driveID, folders := range prevs {
		prevPathCollisions := map[string]string{}

```

File: src/internal/m365/collection/drive/collections.go [hunk 6]
```diff
@@ -72,6 +72,70 @@ func NewCollections(
...
+				// in order to correct the backup state, else we'll potentially hit
+				// unresolvable failures.
+				return map[string]string{}, map[string]map[string]string{}, false, nil
+			}
+
+			prevPathCollisions[prev] = fid
+		}
+
+		// reset the collision check for each drive
+		prevPathCollisions = map[string]string{}
```
Comment (ashmrtn) [inline src/internal/m365/collection/drive/collections.go:133]:
```suggestion
```

File: src/internal/m365/collection/drive/collections_test.go [hunk 7]
```diff
@@ -2839,6 +2910,62 @@ func (suite *OneDriveCollectionsUnitSuite) TestGet() {
...
+					},
+				},
+			},
+			canUsePreviousBackup: false,
+			errCheck:             assert.NoError,
+			prevFolderPaths: map[string]map[string]string{
+				driveID1: {
+					"root":    rootFolderPath1,
+					"folder":  rootFolderPath1 + "/folder",
+					"folder2": rootFolderPath1 + "/folder",
```
Comment (ashmrtn) [inline src/internal/m365/collection/drive/collections_test.go:2938]: could you please add another folder that doesn't appear in the result set and make sure it's marked as deleted

File: src/internal/m365/collection/drive/collections.go [hunk 8]
```diff
@@ -72,6 +72,70 @@ func NewCollections(
...
+						"collision_folder_id_1", fid,
+						"collision_folder_id_2", otherID,
+						"collision_drive_id", driveID,
+						"collision_prev_path", prev).
+					Info("duplicate previous paths across different folder IDs")
+
+				// If a previous path collision occurs, we need to force a full backup
+				// in order to correct the backup state, else we'll potentially hit
+				// unresolvable failures.
+				return map[string]string{}, map[string]map[string]string{}, false, nil
```
Comment (ashmrtn) [inline src/internal/m365/collection/drive/collections.go:126]: This is actually a bit over-conservative. We can simply drop the previous paths and delta for the offending drive instead of for all drives

File: src/internal/m365/collection/drive/collections_test.go [hunk 9]
```diff
@@ -1375,6 +1375,77 @@ func (suite *OneDriveCollectionsUnitSuite) TestDeserializeMetadata() {
...
+					}
+				},
+			},
+			expectedDeltas:       map[string]string{},
+			expectedPaths:        map[string]map[string]string{},
+			canUsePreviousBackup: false,
+			errCheck:             assert.NoError,
+		},
+		{
+			name: "DuplicatePreviousPaths_separateDrives",
```
Comment (ashmrtn) [inline src/internal/m365/collection/drive/collections_test.go:1403]: could you please also add a test where one drive has duplicate paths and the other doesn't? We should only be dropping metadata info for the one that has the duplicate in the ideal case

File: src/internal/m365/collection/drive/collections_test.go [hunk 10]
```diff
@@ -2839,6 +2910,62 @@ func (suite *OneDriveCollectionsUnitSuite) TestGet() {
 				rootFolderPath2: true,
 			},
 		},
+		{
+			name:   "duplicate previous paths in metadata",
```
Comment (ashmrtn) [inline src/internal/m365/collection/drive/collections_test.go:2914]: I'm unsure if we should add another test (or expand this one) to have two drives, one of which has duplicate paths. I think it would be good coverage to have, though of course does come with more maintenance requirements

File: src/internal/m365/collection/drive/collections.go (hunk 1 above)
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections.go:81]: fair, might as well

File: src/internal/m365/collection/drive/collections.go (hunk 2 above)
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections.go:95]: sure, sounds good

File: src/internal/m365/collection/drive/collections.go (hunk 3 above)
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections.go:120]: good catch, thank you

File: src/internal/m365/collection/drive/collections.go (hunk 5 above)
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections.go:112]: Heh, yeah, I already did that in the other PR.   Had a "why did I even do this?" moment.

File: src/internal/m365/collection/drive/collections.go (hunk 8 above)
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections.go:126]: fair point

File: src/internal/m365/collection/drive/collections_test.go (hunk 7 above)
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_test.go:2938]: Hmm, well, it's not marked as deleted.  It's not included in the result set either.  I'll follow up on this in the next PR where we don't force a full enumeration and make sure it's handled the right way there.

File: src/internal/m365/collection/drive/collections_test.go (hunk 9 above)
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_test.go:1403]: Going to skip this.  Same reason as skipping the per-drive metadata removal; we won't be forcing full enumeration on the next PR, so it won't be handled this way.

File: src/internal/m365/collection/drive/collections_test.go (hunk 10 above)
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_test.go:2914]: Sure, that's easy to do.

**Comment by ryanfkeepers** [comment]
@ashmrtn ready for final review

File: src/internal/m365/collection/drive/collections_test.go (hunk 9 above)
Comment (ashmrtn) [reply src/internal/m365/collection/drive/collections_test.go:1403]: I think it's still worth adding a test for this so we can, at the very least, build on it later on. Adding it will help ensure we don't lose track of that case and that alerts act as expected in the next PR

File: src/internal/m365/collection/drive/collections_test.go (hunk 7 above)
Comment (ashmrtn) [reply src/internal/m365/collection/drive/collections_test.go:2952]: I'm not sure if that's correct behavior or not. It would depend on what other data is returned. If we don't return a flag saying the base shouldn't be merged (which we still need to remove at some point) or mark the collection as deleted then it'll be pulled forward

I know this PR train isn't trying to solve that problem anymore, but I'm concerned about losing track of this particular test. It's important to have because we should handle it properly eventually, I just don't know the best way to record it

**Review by ashmrtn — APPROVED** [review]
would really like to see a few more tests added so we don't lose track of them

File: src/internal/m365/collection/drive/collections_test.go (hunk 9 above)
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_test.go:1403]: fwiw, there's already a test for that in the next PR

File: src/internal/m365/collection/drive/collections_test.go (hunk 7 above)
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_test.go:2952]: It's tracked on a test in the next PR.
